ls1intum/Artemis
# [Feature] Highlight code line changes

### Checklist

### Motivation and Context


This PR introduces a diff view for the assessment of programming exercise. Changed or new lines are highlighted with a green background. To calculate the diff we are using `diff-match-patch`, see: https://github.com/google/diff-match-patch . For files that are not in the template (added files during student submission) all lines in the code editor are highlighted.  

### Description

- `diff-match-patch` to calculate the diff
- Add new endpoint to fetch all files with their content 
- New event emitter for `CodeEditorTutorAssessmentComponent` after the file was loaded in the `CodeEditorAceComponent`

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create new programming exercise with manual assessment
4. Participate in exercise with different users and create different submissions (e.g. Submission1 has added in `BubbleSort.java` 3 new lines of code, in `MergeSort.java` added 1 new line of code. Submission2 has appended the comment in `BubbleSort.java` and added a completely new file (e.g. `SortStrategy.java`)
5. Assess programming exercise and check if diff for files are highlighted correctly
6. Manual assessment should work as always, check also students view

### Test Coverage



<img width="919" alt="image" src="https://user-images.githubusercontent.com/41108487/102450314-5f869800-4036-11eb-8cb1-e09660244efd.png">


### Screenshots


Appended comment and added new line:
<img width="1661" alt="image" src="https://user-images.githubusercontent.com/41108487/102446783-dd46a580-402e-11eb-941e-a156e844b69d.png">

New file:
<img width="1659" alt="image" src="https://user-images.githubusercontent.com/41108487/102446823-f51e2980-402e-11eb-8c46-5b3284833a3d.png">

New line and build failed:
<img width="1645" alt="image" src="https://user-images.githubusercontent.com/41108487/102446934-30b8f380-402f-11eb-827b-0642564677cc.png">
